### DISCUSSION TITLE
compile notebook to binary or app to minimise startup time

### DISCUSSION THREAD

**Discussion by gszep** [fonsp/Pluto.jl#810] [Questions about using Pluto]
Would this be possible using `PackageCompiler.jl`?
https://julialang.github.io/PackageCompiler.jl/dev/

**Comment by fonsp** [comment]
Yes, use the notebook file as a representative workload to generate a custom sysimage, then launch Pluto with that sysimage set for notebooks:
```julia
Pluto.run(sysimage="path/to/sysimage.so")
```

**Reply by gszep** [reply]
where do I find the available kwargs for `Pluto.run`?

**Reply by fonsp** [reply]
run `Pluto.run(asdfasdfasdf=123)`

**Reply by stillyslalom** [reply]
Surely it would be better to list these in the `Pluto.run` docstring rather than using an error message for the same purpose

**Reply by gszep** [reply]
how does this kwarg work with Pluto's new built-in package management? The same way sysimages work with pkg?

**Reply by gszep** [reply]
I'm currently following instructions in https://andreaskroepelin.de/blog/plutoslides/ to create a presentation using Pluto and wondering what the best way to reduce startup time is :)

**Reply by ctrekker** [reply]
The best way I know of at the moment to create a sysimage for a Pluto notebook, which will bring significant speedups, is to
1. Create a Julia project with Project.toml and Manifest.toml copied from the notebook
2. Instantiate the project (and add Markdown / InteractiveUtils as well if you use them)
3. Start a julia process inside the project, and when making a system image with PackageCompiler, pass in the path to your notebook for the `precompile_execution_file` kwarg of `create_sysimage`.
4. When starting Pluto.jl, pass the newly generated system image's path into Pluto's `sysimage` kwarg

I've been using this procedure for a little while now and it seems to work fairly well. There is still a bit of overhead from compiling Pluto's `PlutoRunner` module, but as far as I can tell there's no great way to precompile this module until #1881 gets merged. There's also no guarentees this sysimage will behave nicely when running other notebooks, since the Pkg environment could be different.

**Reply by fonsp** [reply]
> how does this kwarg work with Pluto's new built-in package management? The same way sysimages work with pkg?

Exactly, so not very well, but take a look at https://github.com/fonsp/Pluto.jl/pull/1631 and https://github.com/JuliaLang/Pkg.jl/pull/3002 and https://github.com/JuliaLang/julia/pull/44329

**Comment by jcbritobr** [comment]
Its not working anymore. Loading the notebook will make pluto crash

schedule: Task not runnable
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:35
  [2] enq_work(t::Task)
    @ Base .\task.jl:718
  [3] schedule
    @ .\task.jl:753 [inlined]
  [4] get_workspace(session_notebook::Tuple{Pluto.ServerSession, Pluto.Notebook}; allow_creation::Bool)
    @ Pluto.WorkspaceManager C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\evaluation\WorkspaceManager.jl:352

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ffa17b8c611 -- ijl_subtype_env at C:/workdir/src\subtype.c:1854
in expression starting at none:1
ijl_subtype_env at C:/workdir/src\subtype.c:1854
#send_notebook_changes!#388 at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\Dynamic.jl:186
send_notebook_changes! at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\Dynamic.jl:180 [inlined]
response_update_notebook at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\Dynamic.jl:283
jfptr_response_update_notebook_123842 at C:\Users\jcbri\Documents\Workspace\julia\linearalgebra\linearalgebra.dll (unknown line)
process_ws_message at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\WebServer.jl:399
#443 at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\WebServer.jl:190
#upgrade#13 at C:\Users\jcbri\.julia\packages\HTTP\z8l0i\src\WebSockets.jl:439
upgrade at C:\Users\jcbri\.julia\packages\HTTP\z8l0i\src\WebSockets.jl:419 [inlined]
#442 at C:\Users\jcbri\.julia\packages\Pluto\ZBojR\src\webserver\WebServer.jl:171
unknown function (ip: 000001c07935e0a6)
jl_apply at C:/workdir/src\julia.h:1843 [inlined]
jl_f__call_latest at C:/workdir/src\builtins.c:774
#invokelatest#2 at .\essentials.jl:729 [inlined]
invokelatest at .\essentials.jl:726 [inlined]
handle_connection at C:\Users\jcbri\.julia\packages\HTTP\z8l0i\src\Servers.jl:447
#16 at .\task.jl:484
unknown function (ip: 000001c0793435c3)
jl_apply at C:/workdir/src\julia.h:1843 [inlined]
start_task at C:/workdir/src\task.c:931
Allocations: 38827787 (Pool: 38810303; Big: 17484); GC: 46
  [5]
